src/main/scala/com/fulcrumgenomics/umi/DuplexConsensusCaller.scala

```diff
@@ -0,0 +1,255 @@
+import htsjdk.samtools.SAMRecord
+
+/**
+  * Container for constant values and types used by the [[DuplexConsensusCaller]]
+  */
+object DuplexConsensusCaller {
+  /** Various default values for the consensus caller. */
+  val ErrorRatePreUmi: PhredScore         = 45.toByte
+  val ErrorRatePostUmi: PhredScore        = 40.toByte
+  val MinInputBaseQuality: PhredScore     = 15.toByte
+  val NoCall: Byte = 'N'.toByte
+
+  /**
+    * Stores information about a consensus read.  Bases, arrays, and the two single
+    * strand consensus reads must all be the same length.
+    *
+    * @param bases the base calls of the consensus read
+    * @param quals the calculated phred-scaled quality scores of the bases
+    * @param abConsensus the AB consensus read from which the duplex was constructed
+    * @param baConsensus the AB consensus read from which the duplex was constructed
+    */
+  case class DuplexConsensusRead(id: String,
+                                 bases: Array[Byte],
+                                 quals: Array[Byte],
+                                 abConsensus: VanillaConsensusRead,
+                                 baConsensus: VanillaConsensusRead) extends ConsensusRead {
+    require(bases.length == quals.length,  "Bases and qualities are not the same length.")
+    require(bases.length == abConsensus.length, "Bases and AB consensus are not the same length.")
+    require(bases.length == baConsensus.length, "Bases and BA consensus are not the same length.")
+  }
+}
+
+
+/**
+  * Creates duplex consensus reads from SAMRecords that have been grouped by their source molecule
+  * but not yet by source strand.
+  *
+  * Filters incoming bases by quality before building the duplex.
+  *
+  * Output reads and bases are constructed only if there is at least one read from each source
+  * molecule strand.  Otherwise no filtering is performed.
+  *
+  * @param readNamePrefix the prefix to apply to all consensus read names
+  * @param readGroupId    the read group ID to apply to all created consensus reads
+  * @param minInputBaseQuality the minimum input base quality score to use a raw read's base
+  * @param errorRatePreUmi the estimated rate of errors in the DNA prior to attaching UMIs
+  * @param errorRatePostUmi the estimated rate of errors in the DNA post attaching UMIs
+  */
+class DuplexConsensusCaller(override val readNamePrefix: String,
+                            override val readGroupId: String = "A",
+                            val minInputBaseQuality: PhredScore = DuplexConsensusCaller.MinInputBaseQuality,
+                            val errorRatePreUmi: PhredScore = DuplexConsensusCaller.ErrorRatePreUmi,
+                            val errorRatePostUmi: PhredScore = DuplexConsensusCaller.ErrorRatePostUmi
+                           ) extends UmiConsensusCaller[DuplexConsensusRead] {
+
+  private val ssCaller =
+    new VanillaUmiConsensusCaller(readNamePrefix="x", options=new VanillaUmiConsensusCallerOptions(
```

Comment:
Let's chat about this.  I think it's necessary to have a read name prefix, to ensure that the read names in the output BAM have a reasonable chance of being unique.  Otherwise if you generate two BAMs from two libraries, and decide you want to merge them for variant calling, all hell will break loose when you have multiple templates with the same name.